### DISCUSSION TITLE
Nifti Inversion for ResizeWithPadOrCrop shows error

### DISCUSSION THREAD

**Discussion by Rubaida001** [Project-MONAI/MONAI#8037] [General]
Hi there,

I am currently working on harmonizing T1w MRI images. For the preprocessing step, I utilized the ResizeWithPadOrCrop function to resize the data. Following the harmonization process, I tried to convert the inferred image into subject space. I have attempted to use both the **inverse** and **InvertD** functions for this purpose, but unfortunately, none worked. I have checked multiple tutorials but none of the scripts run the **InvertD** function on previously saved datasets.  Could you please help me to solve this problem?

The general steps I am following are as follows:

- Preprocess the data (resize the data into (192,192,192))
- Save the resized images and pass them to the harmonisation model
- Get the harmonized data and run the inversion function to get the original size

Here is the code with **InvertD** function for **forward transform**:

```
load_transformer =  LoadImaged(keys=["image"], image_only=False, ensure_channel_first=True)
    load_image = load_transformer({"image": image_path})
   
 # Define the forward transform pipeline
    preprocess_transform = Compose([
        ResizeWithPadOrCropd(keys=["image"], spatial_size=(192, 192, 192)),
        ToTensord(keys=["image"])
    ])
```

    # Apply the forward transform
    resized_img = preprocess_transform(load_image)
    
    #Save Image    
    img_saved = SaveImage(output_dir=output_dir, resample=False, output_ext="nii", separate_folder=False)
    img_saved(resized_img["image"])

    #Save transformed function and loaded image
    with open('loaded_image.pkl', 'wb') as f:
        pickle.dump(load_image,f)	
    with open('preprocess_transform.pkl', 'wb') as f:
        pickle.dump(preprocess_transform,f)	`

**Inverse transform**:
  ```
#Load pre-process transform function and original image with metadata
  with open('loaded_image.pkl', 'rb') as f:
        batch_data = pickle.load(f)	
   
   with open('preprocess_transform.pkl', 'rb') as f:
        preprocess_transform = pickle.load(f)	
    
#Load the new harmonized image and append it to the original image dictionary
    load_transformer =  LoadImaged(keys=["pred"], image_only=False, ensure_channel_first=True)
    pred_image = load_transformer({"pred":'n4_PSYD_0106_S001_B_T1w_Skl_MNI_scaled_harm.nii'})
    print(pred_image.keys())
    batch_data["pred"] = pred_image["pred"]
```

    # Define the inverse transform pipeline
    inverse_transform = Compose([
        Invertd(
        keys="pred",
        transform=preprocess_transform,
        orig_keys="image",
        meta_keys="image_meta_dict",
        allow_missing_keys=True
             )
    ])
    # Apply the inverse transform
    inverted_data = inverse_transform(batch_data)

But I am getting following error:

`Traceback (most recent call last):
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 141, in apply_transform
    return _apply_transform(transform, data, unpack_items, lazy, overrides, log_stats)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 98, in _apply_transform
    return transform(data, lazy=lazy) if isinstance(transform, LazyTrait) else transform(data)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/utility/dictionary.py", line 464, in inverse
    self.pop_transform(d, key)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/inverse.py", line 338, in pop_transform
    return self.get_most_recent_transform(data, key, check, pop=True)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/inverse.py", line 320, in get_most_recent_transform
    self.check_transforms_match(all_transforms[-1])
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 141, in apply_transform
    return _apply_transform(transform, data, unpack_items, lazy, overrides, log_stats)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 98, in _apply_transform
    return transform(data, lazy=lazy) if isinstance(transform, LazyTrait) else transform(data)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/post/dictionary.py", line 715, in __call__
    inverted = self.transform.inverse(input_dict)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/compose.py", line 364, in inverse
    data = apply_transform(
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 171, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <bound method ToTensord.inverse of <monai.transforms.utility.dictionary.ToTensord object at 0x7fe456d95a00>>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/data/project/PSY_SHARED/HARMONISATION_PROCESSING/CycleGAN/Python/cyclegan/preprocessing/inverse-1.py", line 97, in <module>
    inverse_transform(output_dir)
  File "/data/project/PSY_SHARED/HARMONISATION_PROCESSING/CycleGAN/Python/cyclegan/preprocessing/inverse-1.py", line 76, in inverse_transform
    inverted_data = inverse_transform(batch_data)
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/compose.py", line 335, in __call__
    result = execute_compose(
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/compose.py", line 111, in execute_compose
    data = apply_transform(
  File "/home/k2031007/.local/lib/python3.9/site-packages/monai/transforms/transform.py", line 171, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.post.dictionary.Invertd object at 0x7fe26c743af0>
`

**Comment by KumoLiu** [comment]
Hi @Rubaida001, 
Looks like you want to apply invert on a new data without any information, perhaps you can try include the meta information from existing data. Something like:
```
batch_data["pred"] = MetaTensor(pred_image["pred"], meta=exsiting_data.meta)
```
You can refer to this tutorial for more details about how to using invert: https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/spleen_segmentation_3d.ipynb

Hope it helps, thanks.

**Reply by Rubaida001** [reply]
Thank you so much for the suggestion. I changed the code and included meta information from ** resized_img["image_meta_dict"]** (output of forward transforms). But the error remained same.

`batch_data["pred"] = MetaTensor(pred_image["pred"], meta=resized_img["image_meta_dict"])`

Just for the clarification:
- Original MRI data size: 182, 218, 182
- After forward transforms: 192, 192, 192 (Run harmonization model on those data)
- Now I want to go back to the original space: 182, 218, 182

**Reply by Rubaida001** [reply]
Hi,
I followed the tutorial and given more details regarding the predicted data. The updated inverse transform is the following: 
    
   ```
 load_transformer =  LoadImaged(keys=["pred"], image_only=False, ensure_channel_first=True)
    pred_image = load_transformer({"pred":predicted.nii})
       
    batch_data["pred"] = pred_image["pred"] 
    batch_data["pred_meta_dict"] = pred_image["pred_meta_dict"]    

    # Define the inverse transform pipeline
    inverse_transform = Compose([
        Invertd(
        keys="pred",
        transform=preprocess_transform,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        nearest_interp=False,
        to_tensor=True,
        allow_missing_keys=True)])
    inverted_data = inverse_transform(batch_data)
```
However, I am still getting the same error (**get_most_recent_transform self.check_transforms_match(all_transforms[-1])  IndexError: list index out of range**). 

Any suggestions would be appreciated. Thank you.

**Reply by KumoLiu** [reply]
Please check whether the resized_img["image_meta_dict"] has `applied_operations`  by using `resized_img["image_meta_dict"]["applied_operations"]`, By the way, since 0.9.1 we have introduced MetaTensor, you can direct use `resized_img.applied_operations` when data is a MetaTensor.
https://github.com/Project-MONAI/MONAI/wiki/MetaTensor-guide

**Reply by Rubaida001** [reply]
`resized_img` is in MetaTensor format; however, there are no `applied_operations` in the keys. I am using `LoadImaged` to read the NIfti files. I am not sure if I should do it in a different way.

**Reply by KumoLiu** [reply]
If there is no `applied_operations` in the data, then you can't apply invert. You need just ensure the `preprocess_transform` have been applied on  `resized_img`, then you can use invert.
I would suggest you play with this notebook to see what will be saved in the `applied_operations` in the images.
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/spleen_segmentation_3d.ipynb
